### PR TITLE
[CAM-10027] feat(assert): Add hasBusinessKey assertion

### PR DESCRIPTION
**Pull request by dlavrenuek**
- implements hasBusinessKey process instance assertion
- adds documentation on hasBusinessKey process instance assertion

related to CAM-10027

### CODE REVIEW

**Comment by dlavrenuek** [comment]
This PR also has a small typo fix in the documentation:

> You will e.g. often have **a a** single open task instance

**Comment by yanavasileva** [comment]
Hi @dlavrenuek,

Thank you for the contribution!
I will provide you feedback most probably next week.

Best regards,
Yana

File: core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * @author Dimitri Lavrenük (dimitri.lavr@gmail.com)
+ */
+public class ProcessInstanceAssertHasBusinessKeyTest extends ProcessAssertTestCase {
+
+  @Rule
+  public ProcessEngineRule processEngineRule = new ProcessEngineRule();
+
+  @Test
+  @Deployment(resources = {"bpmn/ProcessInstanceAssert-hasBusinessKey.bpmn"})
```
Comment (yanavasileva) [inline core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java:38]: I would propose to not create a new process model but reuse one of the existing, e.g. `ProcessInstanceAssert-isActive.bpmn`. As the process doesn't need any special design.

File: core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import org.camunda.bpm.engine.test.assertions.helpers.Failure;
+import org.camunda.bpm.engine.test.assertions.helpers.ProcessAssertTestCase;
+import org.junit.Rule;
+import org.junit.Test;
+
+import static org.camunda.bpm.engine.test.assertions.ProcessEngineTests.*;
+
+/**
+ * @author Dimitri Lavrenük (dimitri.lavr@gmail.com)
+ */
```
Comment (yanavasileva) [inline core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java:31]: The `author` tags are no longer necessary in new contributions as git tracks in good manner the author of the code and the author tag in javadoc seems only to clutter the code. (However, we don't remove an existing tags.) So I would suggest to remove it if you are fine with that.

**Review by yanavasileva — CHANGES_REQUESTED** [review]
The contribution looks good. I have only small suggestions to improve the test.

File: core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java (hunk 1 above)
Comment (dlavrenuek) [reply core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java:38]: I did this according to other tests that each have a separate process model definition file while most of them are exactly the same except for the process id. I found 25 of those files, some exmaples:
- ProcessInstanceAssert-isActive.bpmn
- ProcessInstanceAssert-isEnded.bpmn
- ProcessInstanceAssert-isStarted.bpmn
- ProcessInstanceAssert-hasVariables.bpmn
- ProcessInstanceAssert-hasNoVariables.bpmn

If I would replace it here with an existing file, the other tests should also be refactored, right?

File: core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  @Test
+  @Deployment(resources = {"bpmn/ProcessInstanceAssert-hasBusinessKey.bpmn"})
+  public void testHasProcessDefinitionKey_Failure() {
+    // Given
+    final ProcessInstance processInstance = runtimeService().startProcessInstanceByKey(
+      "ProcessInstanceAssert-hasBusinessKey",
+      "anotherBusinessKey"
+    );
+    // When
+    runtimeService().suspendProcessInstanceById(processInstance.getId());
```
[the remark this answers is not in the document]
Comment (dlavrenuek) [reply core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java]: Fixed 👍

File: core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java (hunk 2 above)
Comment (dlavrenuek) [reply core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java]: I am totally fine with that, removed it.

File: core/src/main/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssert.java [hunk 4]
```diff
@@ -373,6 +373,24 @@ public ProcessInstanceAssert hasProcessDefinitionKey(String processDefinitionKey
...
+   * Verifies the expectation that the {@link ProcessInstance} has the
+   * given businessKey.
+   *
+   * @param businessKey the expected key
+   * @return  this {@link ProcessInstanceAssert}
+   */
+  public ProcessInstanceAssert hasBusinessKey(String businessKey) {
+    isNotNull();
+    final String message = "Expecting %s to have business key '%s', " +
+      "but instead found it to have business key key '%s'.";
```
Comment (tmetzke) [inline core/src/main/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssert.java:386]: There is one extra "key" here, right?
```suggestion
      "but instead found it to have business key '%s'.";
```

File: core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java (hunk 1 above)
Comment (tmetzke) [reply core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java:35]: I guess both options are fine here, adding a new one is also OK if you already did it and it also separates the concerns well. You don't necessarily have to add one, but it's also fine to add one, I would say. 😃

File: core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import static org.camunda.bpm.engine.test.assertions.ProcessEngineTests.*;
+
+public class ProcessInstanceAssertHasBusinessKeyTest extends ProcessAssertTestCase {
+
+  @Rule
+  public ProcessEngineRule processEngineRule = new ProcessEngineRule();
+
+  @Test
+  @Deployment(resources = {"bpmn/ProcessInstanceAssert-hasBusinessKey.bpmn"})
+  public void testHasProcessDefinitionKey_Success_With_String() {
```
Comment (tmetzke) [inline core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java:36]: Test method names are a copy-paste mistake, right?

File: core/src/main/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssert.java (hunk 4 above)
Comment (dlavrenuek) [reply core/src/main/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssert.java]: Thanks, fixed it

File: core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java (hunk 5 above)
Comment (dlavrenuek) [reply core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java]: You are right 👍 fixed

File: core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java (hunk 1 above)
Comment (yanavasileva) [reply core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java:35]: I agree with Tobi here.

Comment (dlavrenuek) [reply core/src/test/java/org/camunda/bpm/engine/test/assertions/bpmn/ProcessInstanceAssertHasBusinessKeyTest.java:35]: @yanavasileva does that mean that it is fine with the the `ProcessInstanceAssert-hasBusinessKey.bpmn` file or that existing test suits should be refactored to only use one single definition if it is the same?

**Review by yanavasileva — APPROVED** [review]

**Comment by yanavasileva** [comment]
@dlavrenuek, thank you for implementing the review hints and for the contribution once again.

**Pull request merged**